### ISSUE TITLE
Question about supporting APL Authoring Tool in public beta release

### ISSUE THREAD

**Issue by kenbrower** [timheuer/alexa-skills-dotnet#159]
We have been using APL with the standard body templates and we can add them to a response using the Response.Directives.Add just fine. I have recently been looking at the new APL Authoring Tool which is currently in public beta release (see Display option under the Build tab in the Skill developer console) or go here https://developer.amazon.com/alexa/console/ask/displays/.

The problem is that I can't figure out how to use the Data JSON from the authoring tool to create a valid APL directive. When this JSON is used to try and add a directive we receive an exception saying something like "value cannot be null".  What am I missing about how to use this new authoring tool to build APL directives?  Do you need to save a created template into your skill somehow before you can send related data in a directive?

Any information or reference to related Amazon documentation would be most appreciated.
Thanks!

**Comment by stoiveyp** [comment]
Hi @kenbrower would you mind if I asked for a little clarification? APL and body templates are two different techniques to rendering data on screen; with APL being the preferred mechanism from Amazon these days due to its increased flexibility - but they don't work in conjunction with one another.

If you're using body templates, then the data has to come from filling in the fields in the template and attaching that to the render template directive.

If you're using APL then the main template has a "payload" parameter - and that refers to the data sources tab in the designer, so all your fields will be of the style  `${payload.dataSourceName.fieldName}`

If you want to have support for APL within Alexa.NET skills - then that's available through the Alexa.NET.APL NuGet package at https://github.com/stoiveyp/Alexa.NET.APL and you can use the RenderDocument directive, which has a data sources property in it for your data. More than happy to answer questions about APL support there

I realise this is quite a wordy answer, but I want to make sure we're all talking about the same thing, and that the issues are in the right place to best get resolved for you 👍

**Comment by kenbrower** [comment]
Hi @stoiveyp . Thank you for the reply and the useful information. After reading your message I realize I was mixing the two and in fact what we have been using is just the body templates which I am now assuming was the precursor to APL.

I'll work on setting up a test using the RenderDocument directive and see if we can figure out how to get that working. If you don't mind one more naive question... When you use the Alexa developer authoring tool where is the actual template JSON and how do you send that with the RenderDocument directive?  It makes sense that the data is referenced in a payload parameter but I'm not understanding where the actual template definition resides.

Thanks again for the quick reply!

**Comment by stoiveyp** [comment]
@kenbrower really good question - the tool is only a way of creating the APL templates, to see the template you're creating with the tools there is the toggle button (highlighted below)

![templateButton](https://user-images.githubusercontent.com/147125/60835393-e2a58c80-a1ba-11e9-83da-7090747cbfe1.jpg)

Once you have a template you're happy with - that template needs to be copied/saved out of the designer and stored somewhere the skill can access it (S3 bucket, local file etc). This can then be used to deserialize the JSON into a Document object that can be used with the directive.

Hope this helps.

**Comment by kenbrower** [comment]
Hey @stoiveyp ,

That's the all important toggle button we had overlooked!  Makes complete sense now that I see how to get both the template and the data from the designer.

Thanks for your help. Really enjoying working with the Alexa Skills SDK for .NET.

**Comment by kenbrower** [comment]
Hi @stoiveyp , 

Thanks again for the guidance provided in the thread above. We have included the Alexa.NET.APL NuGet and made some progress on supporting APL.  It seems a lot can change in a day as the Amazon looks to have recently been updated and the toggle button replaced with a left-side menu 🙂.

After creating some APL we are running into problems trying to deserialize it into a RenderDocumentDirective using code like the following.

```
RenderDocumentDirective directive = new RenderDocumentDirective();
directive.Document = JsonConvert.DeserializeObject<APLDocument>(apl);
```

Most of our errors seem to be coming from the onMount and/or the commands under the onMount. Not sure if something has changed recently but the specification seems to indicate that onMount is an array containing objects that would have type, commands, finally, etc. as children. We get exceptions like the following when using what looks to be properly formatted JSON for the onMount value.

`Error setting value to 'OnMount' on 'Alexa.NET.Response.APL.APLDocument'.`

It also seems that it should be acceptable to simple include a blank array for onMount but when we simply try and include a blank array ""onMount": []," we get an exception saying "Value cannot be null. Parameter name: key" during the deserialization.

Should we be able to deserialize from APL taken from the authoring tool and create a RenderDocumentDirective object?

Thanks.

**Comment by stoiveyp** [comment]
@kenbrower as this an issue specifically with the APL NuGet package - would you mind closing this issue and raising it there?

https://github.com/stoiveyp/Alexa.NET.APL

both null and empty arrays should have been fine, so I'll start taking a look in the meantime

**Comment by stoiveyp** [comment]
@kenbrower thanks for finding this - issue with the converter, shouldn't code so late!

it's been fixed in v4.0.2 https://www.nuget.org/packages/Alexa.NET.APL/4.0.2

**Comment by kenbrower** [comment]
Hi @stoiveyp!   Impressive turnaround, thanks so much.  We'll pull down the latest NuGet version and do some testing with our code.

If we see anything else I'll open a new issue under the Alexa.NET.APL project.
